storage/measure.go

```diff
@@ -0,0 +1,221 @@
+		}
+	}
+	var favoriteData map[string]*FavoriteNode
+	err := util.ReadJSON(s.favoriteFilePath, &favoriteData)
+	if err != nil {
+		return err
+	}
+
+	s.FavoriteNodes = NewStorage()
+	for k, v := range favoriteData {
+		s.FavoriteNodes.Add(k, v)
+	}
+
+	isExists = util.Exists(s.badFilePath)
+	if !isExists {
+		err := util.WriteJSON(s.badFilePath, map[string]*BadNode{})
+		if err != nil {
+			return err
+		}
+	}
+	var badData map[string]BadNodes
+	err = util.ReadJSON(s.badFilePath, &badData)
+	if err != nil {
+		return err
+	}
+	s.BadNodes = badData
+
+	err = s.ClearFavoriteExpired()
+	if err != nil {
+		return err
+	}
+	err = s.ClearBadExpired()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *MeasureStorage) ClearFavoriteExpired() error {
+	for k, v := range s.FavoriteNodes.GetData() {
+		if time.Now().Unix() > v.(*FavoriteNode).ExpiresAt {
+			s.FavoriteNodes.Delete(k)
+		}
+	}
+	err := s.SaveFavoriteNodes()
+	return err
+}
+
+func (s *MeasureStorage) ClearBadExpired() error {
+	s.badNodeMutex.Lock()
+	defer s.badNodeMutex.Unlock()
+	for k1, v1 := range s.BadNodes {
+		for k2, v2 := range v1 {
+			if time.Now().Unix() > v2.ExpiresAt {
+				delete(v1, k2)
+			}
+		}
+		if len(v1) == 0 {
+			delete(s.BadNodes, k1)
+		}
+	}
+	err := s.SaveBadNodes()
+
+	return err
+}
+
+func (s *MeasureStorage) SaveFavoriteNodes() error {
+	s.favoriteNodeFileMutex.Lock()
+	defer s.favoriteNodeFileMutex.Unlock()
+	err := util.WriteJSON(s.favoriteFilePath, s.FavoriteNodes.GetData())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *MeasureStorage) SaveBadNodes() error {
+	s.badNodeFileMutex.Lock()
+	defer s.badNodeFileMutex.Unlock()
+	err := util.WriteJSON(s.badFilePath, s.BadNodes)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *MeasureStorage) AddFavoriteNode(key string, val *FavoriteNode) {
+	if val.ExpiresAt == 0 {
+		val.ExpiresAt = time.Now().Add(favoriteExpired).Unix()
+	}
+
+	if s.FavoriteNodes.Len() >= maxFavoriteLength {
+		minBandwidth := float32(0)
+		for _, v := range s.FavoriteNodes.GetData() {
+			item := v.(*FavoriteNode)
+			if item.MinBandwidth < minBandwidth || minBandwidth == 0 {
+				minBandwidth = item.MinBandwidth
+			}
+		}
+		if val.MinBandwidth > minBandwidth {
+			s.FavoriteNodes.Add(key, val)
+		}
+		deleteKey := ""
+		minExpire := int64(math.MaxInt32)
+		for k, v := range s.FavoriteNodes.GetData() {
+			item := v.(*FavoriteNode)
+			if item.ExpiresAt < minExpire {
+				minExpire = item.ExpiresAt
+				deleteKey = k
+			}
+		}
+		s.FavoriteNodes.Delete(deleteKey)
+	} else {
+		s.FavoriteNodes.Add(key, val)
+	}
+}
+
+func (s *MeasureStorage) AddBadNode(key string, val *BadNode) {
+	s.badNodeMutex.Lock()
+	defer s.badNodeMutex.Unlock()
+	if val.ExpiresAt == 0 {
+		val.ExpiresAt = time.Now().Add(badExpired).Unix()
+	}
+
+	if val.MaskSize == 0 {
+		val.MaskSize = maskSize
+	}
+
+	_, subnet, err := net.ParseCIDR(fmt.Sprintf("%s/%d", key, val.MaskSize))
+	if err != nil {
+		log.Println(err)
```

Comment:
Should we return here? If `err != nil`, `subnet` is probably `nil` and the next part will panic